File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/loaders/ClassFieldLookup.java
package com.github.ol_loginov.heaplibweb.services.loaders;

import com.github.ol_loginov.heaplibweb.repository.EntityIdentity;
import com.github.ol_loginov.heaplibweb.repository.heap.FieldEntity;
import com.github.ol_loginov.heaplibweb.repository.heap.HeapEntity;
import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories;
import lombok.RequiredArgsConstructor;
import org.netbeans.lib.profiler.heap.Field;

import java.util.Map;
import java.util.stream.Collectors;

@RequiredArgsConstructor
class ClassFieldLookup {
	private final HeapEntity heapEntity;
	private final HeapRepositories heapRepositories;
	private final TypeIdLookup typeIdLookup;
	private Map<FieldKey, Integer> fieldEntities;

	private record FieldKey(long declaringClassId, String name, boolean isStatic) {
	}

	public int getFieldEntityId(Field field) {
		if (fieldEntities == null) {
			fieldEntities = heapRepositories.getFields().streamAllByHeapId(heapEntity.getId())
				.collect(Collectors.toMap(
					e -> new FieldKey(e.getDeclaringClassId(), e.getName(), e.isStaticFlag()),
					EntityIdentity::getId
				));
		}
		var fieldKey = new FieldKey(field.getDeclaringClass().getJavaClassId(), field.getName(), field.isStatic());
		return fieldEntities.computeIfAbsent(fieldKey, k -> this.createField(field));
	}

	private int createField(Field field) {
		var fieldEntity = new FieldEntity(heapEntity.getId(),
			field.getDeclaringClass().getJavaClassId(), field.getName(), field.isStatic(),
			typeIdLookup.lookupTypeId(field.getType().getName()));
		heapRepositories.persist(fieldEntity);
		return fieldEntity.getId();
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/loaders/ClassFieldLookup.kt
package com.github.ol_loginov.heaplibweb.services.loaders

import com.github.ol_loginov.heaplibweb.repository.heap.HeapScope
import org.netbeans.lib.profiler.heap.Field
import org.springframework.transaction.support.TransactionOperations
import kotlin.streams.asSequence

internal class ClassFieldLookup(
    private val scope: HeapScope,
    private val transactionOperations: TransactionOperations
) {
    private data class FieldKey(val declaringClassId: Long, val name: String, val isStatic: Boolean)

    private val fieldEntities: MutableMap<FieldKey, Int> by lazy {
        scope.fields.streamAll().use {
            it.asSequence()
                .associateBy({ FieldKey(it.declaringClassId, it.name, it.staticFlag) }, { it.id })
                .toMutableMap()
        }
    }

    fun getFieldEntityId(field: Field): Int {
        val fieldKey = FieldKey(field.declaringClass.javaClassId, field.name, field.isStatic)
        return fieldEntities.computeIfAbsent(fieldKey) { this.createField(field) }
    }

    private fun createField(field: Field): Int {
        TODO("not implemented")
//        return transactionOperations.execute {
//            val fieldEntity = FieldEntity(
//                field.declaringClass.javaClassId, field.name, field.isStatic,
//                typeIdLookup.lookupTypeId(field.type.)
//            )
//            scope.fields.persist(fieldEntity)
//            fieldEntity.id
//        }!!
    }
}